### DISCUSSION TITLE
types.union only supports up to 9 arguments for correct typing result

### DISCUSSION THREAD

**Discussion by limbosounds** [mobxjs/mobx-state-tree#2083] [Q&A] [answered]
Here is an example:
<img width="454" alt="image" src="https://github.com/mobxjs/mobx-state-tree/assets/26549754/e34fcdc0-117e-4ff6-9c9c-08ac7f6e8417">

and when we stripping out 10th model, it is starting to work normally:
<img width="856" alt="image" src="https://github.com/mobxjs/mobx-state-tree/assets/26549754/92a3f93a-f5ee-4862-9ee9-db4ff025d265">

The only way I can see to workaround this is to use union of unions, but it sounds like schizophrenia :)

Am I missing something?

**Comment by coolsoftwaretyler** [comment]
Hey @limbosounds - this sounds like it might be a typescript issue on our end. I'm not sure exactly what's going on there (hopefully someone else does). I'm sorry I don't have a better answer for you, but I'd be happy to move this to the issue tracker if you like!

We are slowly working through a lot of the existing issues, including our overall TypeScript challenges, so hopefully we can figure this out for you in the coming months.

**Reply by limbosounds** [reply, marked as answer]
Hello and thank you for your response!
This issue has a workaround:
```typescript
types.union(
    types.union(
        // ...up to 9 types
    ),
    types.union(
        // ...up to 9 types
    ),
    // ...up to 9 unions
)
```
This workaround supports up to 81 types in total and resulting type is just the same as if it was a plain union.
So, it's kinda suitable for me despite looking schizophrenic 🤣 

I tried to discover the source of this issue, and seems like it is in `types.union` definition
It has 33 overloads, and each next has more generic types than previous
Simplified example
```typescript
declare function union<T1>(model1: T1)
declare function union<T1, T2>(model1: T1, model2: T2)
declare function union<T1, T2, T3>(model1: T1, model2: T2, model3: T3)
// and so on
// but in original types there are more additional helper generics
```
My bet is there are just not enough overloads for unions with 10+ models

Hope this can help you in future)

**Reply by coolsoftwaretyler** [reply]
Thanks @limbosounds! I appreciate the follow up. Very helpful. I'm going to mark that as the answer for now, and hopefully we can improve this for folks in the future.
